### ISSUE TITLE
Serial connection to homeassistant

### ISSUE THREAD

**Issue by liakoumedos** [ParadoxAlarmInterface/pai#148]
Hi, i am trying the addon to my ha, but i am always getting the error:
Traceback (most recent call last):
  File "/usr/local/lib/python3.7/site-packages/serial/serialposix.py", line 265, in open
    self.fd = os.open(self.portstr, os.O_RDWR | os.O_NOCTTY | os.O_NONBLOCK)
PermissionError: [Errno 13] Permission denied: '/dev/ttyUSB0'
During handling of the above exception, another exception occurred:
Traceback (most recent call last):
  File "/usr/local/lib/python3.7/site-packages/paradox/main.py", line 80, in run_loop
    if await alarm.connect():
  File "/usr/local/lib/python3.7/site-packages/paradox/paradox.py", line 94, in connect
    if not await self.connection.connect():
  File "/usr/local/lib/python3.7/site-packages/paradox/connections/serial_connection.py", line 51, in connect
    self.loop, self.make_protocol, self.port_path, self.baud
  File "/usr/local/lib/python3.7/asyncio/coroutines.py", line 120, in coro
    res = func(*args, **kw)
  File "/usr/local/lib/python3.7/site-packages/serial_asyncio/__init__.py", line 410, in create_serial_connection
    ser = serial.serial_for_url(*args, **kwargs)
  File "/usr/local/lib/python3.7/site-packages/serial/__init__.py", line 88, in serial_for_url
    instance.open()
  File "/usr/local/lib/python3.7/site-packages/serial/serialposix.py", line 268, in open
    raise SerialException(msg.errno, "could not open port {}: {}".format(self._port, msg))
serial.serialutil.SerialException: [Errno 13] could not open port /dev/ttyUSB0: [Errno 13] Permission denied: '/dev/ttyUSB0'

**Comment by yozik04** [comment]
```
chmod a+w /dev/ttyUSB0
```
helps?

**Comment by liakoumedos** [comment]
no. still having the same error

**Comment by yozik04** [comment]
What `ls -al /dev/ttyUSB0` shows?

**Comment by ShadowofReason** [comment]
I assume that HA runs under another user and not under “root”.

Is that user a member of “dialout” group?

**Comment by liakoumedos** [comment]
ls -al /dev/ttyUSB0 shows crw-rw--w-     1 root audio 188, 0 Mar 30 22:37 /dev/ttyUSB0

**Comment by yozik04** [comment]
```
chmod a+rw /dev/ttyUSB0
```
You need to allow read as well.

**Comment by liakoumedos** [comment]
no, it still not connecting.
Traceback (most recent call last):
  File "/usr/local/lib/python3.7/site-packages/serial/serialposix.py", line 265, in open
    self.fd = os.open(self.portstr, os.O_RDWR | os.O_NOCTTY | os.O_NONBLOCK)
PermissionError: [Errno 13] Permission denied: '/dev/ttyUSB0'
During handling of the above exception, another exception occurred:
Traceback (most recent call last):
  File "/usr/local/lib/python3.7/site-packages/paradox/main.py", line 80, in run_loop
    if await alarm.connect():
  File "/usr/local/lib/python3.7/site-packages/paradox/paradox.py", line 94, in connect
    if not await self.connection.connect():
  File "/usr/local/lib/python3.7/site-packages/paradox/connections/serial_connection.py", line 51, in connect
    self.loop, self.make_protocol, self.port_path, self.baud
  File "/usr/local/lib/python3.7/asyncio/coroutines.py", line 120, in coro
    res = func(*args, **kw)
  File "/usr/local/lib/python3.7/site-packages/serial_asyncio/__init__.py", line 410, in create_serial_connection
    ser = serial.serial_for_url(*args, **kwargs)
  File "/usr/local/lib/python3.7/site-packages/serial/__init__.py", line 88, in serial_for_url
    instance.open()
  File "/usr/local/lib/python3.7/site-packages/serial/serialposix.py", line 268, in open
    raise SerialException(msg.errno, "could not open port {}: {}".format(self._port, msg))
serial.serialutil.SerialException: [Errno 13] could not open port /dev/ttyUSB0: [Errno 13] Permission denied: '/dev/ttyUSB0

**Comment by yozik04** [comment]
Is it hass.io or what? Do you have access to root os or only ssh container?

**Comment by liakoumedos** [comment]
Hostname | hassio
-- | --
System | HassOS 3.12

**Comment by yozik04** [comment]
Here is my udev rule:
```
$ cat /etc/udev/rules.d/paradox-serial.rules
SUBSYSTEMS=="usb", ATTRS{idVendor}=="0403", ATTRS{idProduct}=="6001", SYMLINK+="paradox", MODE="0666"
```

It needs to be in host os. Not in a container.

I am thinking if PAI is running as root in the container then probably I can try to chmod serial file.

**Comment by yozik04** [comment]
I will try in the evening.

**Comment by liakoumedos** [comment]
thanks

**Comment by yozik04** [comment]
Try latest development version. I do not think it will help. But just a try.

**Comment by liakoumedos** [comment]
i tried the latest dev version with the same result.
2020-03-31 18:23:47,060 - INFO     - PAI - Starting Paradox Alarm Interface 3.0.0-dev
2020-03-31 18:23:47,061 - INFO     - PAI - Config loaded from /data/options.json
2020-03-31 18:23:47,063 - INFO     - PAI - Console Log level set to 20
2020-03-31 18:23:47,168 - INFO     - PAI.paradox.interfaces - Starting BasicMQTTInterface Interface
2020-03-31 18:23:47,171 - INFO     - PAI.paradox.interfaces.mqtt.core - MQTT loop started
2020-03-31 18:23:47,178 - INFO     - PAI.paradox.interfaces - Starting HomeAssistantMQTTInterface Interface
2020-03-31 18:23:47,179 - INFO     - PAI - Starting...
2020-03-31 18:23:47,181 - INFO     - PAI.paradox.paradox - Connecting to interface
2020-03-31 18:23:47,181 - INFO     - PAI.paradox.paradox - Using Serial Connection
2020-03-31 18:23:47,183 - INFO     - PAI.paradox.connections.serial_connection - Connecting to serial port /dev/ttyUSB0
2020-03-31 18:23:47,184 - INFO     - PAI.paradox.connections.serial_connection - /dev/ttyUSB0 is not readable/writable. Trying to fix...
2020-03-31 18:23:47,186 - ERROR    - PAI.paradox.connections.serial_connection - Failed to update file /dev/ttyUSB0 permissions
2020-03-31 18:23:47,188 - ERROR    - PAI - PAI Critical exception. Stopping PAI
Traceback (most recent call last):
  File "/usr/local/lib/python3.7/site-packages/serial/serialposix.py", line 265, in open
    self.fd = os.open(self.portstr, os.O_RDWR | os.O_NOCTTY | os.O_NONBLOCK)
PermissionError: [Errno 13] Permission denied: '/dev/ttyUSB0'
During handling of the above exception, another exception occurred:
Traceback (most recent call last):
  File "/usr/local/lib/python3.7/site-packages/paradox/connections/serial_connection.py", line 76, in connect
    self.loop, self.make_protocol, self.port_path, self.baud
  File "/usr/local/lib/python3.7/asyncio/coroutines.py", line 120, in coro
    res = func(*args, **kw)
  File "/usr/local/lib/python3.7/site-packages/serial_asyncio/__init__.py", line 410, in create_serial_connection
    ser = serial.serial_for_url(*args, **kwargs)
  File "/usr/local/lib/python3.7/site-packages/serial/__init__.py", line 88, in serial_for_url
    instance.open()
  File "/usr/local/lib/python3.7/site-packages/serial/serialposix.py", line 268, in open
    raise SerialException(msg.errno, "could not open port {}: {}".format(self._port, msg))
serial.serialutil.SerialException: [Errno 13] could not open port /dev/ttyUSB0: [Errno 13] Permission denied: '/dev/ttyUSB0'
The above exception was the direct cause of the following exception:
Traceback (most recent call last):
  File "/usr/local/lib/python3.7/site-packages/paradox/main.py", line 79, in run_loop
    if await alarm.connect():
  File "/usr/local/lib/python3.7/site-packages/paradox/paradox.py", line 102, in connect
    if not await self.connection.connect():
  File "/usr/local/lib/python3.7/site-packages/paradox/connections/serial_connection.py", line 80, in connect
    raise SerialConnectionOpenFailed("Connection to serial port failed") from e
paradox.exceptions.SerialConnectionOpenFailed: Connection to serial port failed
2020-03-31 18:23:47,195 - INFO     - PAI.paradox.paradox - Disconnecting from the Alarm Panel
2020-03-31 18:23:47,196 - INFO     - PAI.paradox.paradox - Clean Session
2020-03-31 18:23:47,197 - INFO     - PAI.paradox.interfaces.interface_manager - Stopping all interfaces
2020-03-31 18:23:47,198 - INFO     - PAI.paradox.interfaces.interface_manager - Stopping BasicMQTTInterface
2020-03-31 18:23:47,200 - INFO     - PAI.paradox.interfaces.interface_manager - Stopping HomeAssistantMQTTInterface
2020-03-31 18:23:48,172 - INFO     - PAI.paradox.interfaces.mqtt.core - MQTT loop stopped
2020-03-31 18:23:48,173 - INFO     - PAI - Good bye!
Attempting to load configuration from /data/options.json

**Comment by yozik04** [comment]
Then I am currently out of ideas. Hassio is really closed system and you do not have access to root os where actual permission changes should help.

**Comment by liakoumedos** [comment]
yes something is happening with the rules, but from hassio i dont have access.

**Comment by juanbrunette** [comment]
So this is how I hacked this error: 
serial.serialutil.SerialException: [Errno 13] could not open port /dev/ttyUSB0: [Errno 13] Permission denied: '/dev/ttyUSB0'

1) Connect a screen and keybard to your Raspberry PI running HassOS.
2) Press Enter
    - You will see homeassistant login
3) Login as root
    - You will see ha >
4) Type in: login
    - You will see # (this means you are now on the HassOs Terminal as root)
5) Type in: vi /etc/udev/rules.dev/paradox.rules
6) Insert the following into the file by pressing i on your keyboard: SUBSYSTEMS=="usb", ATTRS{idVendor}=="0403", ATTRS{idProduct}=="6001", SYMLINK+="paradox", MODE="0666"
7) Press Esc
8) Type in (note the : in front of wq): :wq
9) Type in: cat paradox.rules
    - You should see SUBSYSTEMS=="usb", ATTRS{idVendor}=="0403", ATTRS{idProduct}=="6001", SYMLINK+="paradox", MODE="0666" being displayed 
10) Reboot and your Paradox Addon should now be able to access /dev/ttyUSB0

**Comment by juanbrunette** [comment]
> Then I am currently out of ideas. Hassio is really closed system and you do not have access to root os where actual permission changes should help.

So this is how I hacked this error: 
serial.serialutil.SerialException: [Errno 13] could not open port /dev/ttyUSB0: [Errno 13] Permission denied: '/dev/ttyUSB0'

1) Connect a screen and keybard to your Raspberry PI running HassOS.
2) Press Enter
    - You will see homeassistant login
3) Login as root
    - You will see ha >
4) Type in: login
    - You will see # (this means you are now on the HassOs Terminal as root)
5) Type in: vi /etc/udev/rules.dev/paradox.rules
6) Insert the following into the file by pressing i on your keyboard: SUBSYSTEMS=="usb", ATTRS{idVendor}=="0403", ATTRS{idProduct}=="6001", SYMLINK+="paradox", MODE="0666"
7) Press Esc
8) Type in (note the : in front of wq): :wq
9) Type in: cat paradox.rules
    - You should see SUBSYSTEMS=="usb", ATTRS{idVendor}=="0403", ATTRS{idProduct}=="6001", SYMLINK+="paradox", MODE="0666" being displayed 
10) Reboot and your Paradox Addon should now be able to access /dev/ttyUSB0

**Comment by juanbrunette** [comment]
> yes something is happening with the rules, but from hassio i dont have access.

So this is how I hacked this error: 
serial.serialutil.SerialException: [Errno 13] could not open port /dev/ttyUSB0: [Errno 13] Permission denied: '/dev/ttyUSB0'

1) Connect a screen and keybard to your Raspberry PI running HassOS.
2) Press Enter
    - You will see homeassistant login
3) Login as root
    - You will see ha >
4) Type in: login
    - You will see # (this means you are now on the HassOs Terminal as root)
5) Type in: vi /etc/udev/rules.dev/paradox.rules
6) Insert the following into the file by pressing i on your keyboard: SUBSYSTEMS=="usb", ATTRS{idVendor}=="0403", ATTRS{idProduct}=="6001", SYMLINK+="paradox", MODE="0666"
7) Press Esc
8) Type in (note the : in front of wq): :wq
9) Type in: cat paradox.rules
    - You should see SUBSYSTEMS=="usb", ATTRS{idVendor}=="0403", ATTRS{idProduct}=="6001", SYMLINK+="paradox", MODE="0666" being displayed 
10) Reboot and your Paradox Addon should now be able to access /dev/ttyUSB0

**Comment by B4b4sh** [comment]
Hi,
I have same problems. [Errno 13] Permission denied.
I tried all solutions proposed here, but unfortunately still not working.
I work with a VirtualBox HomeAssistant and my ParadoxUSB307 serial.
In HA the serial is well recognized because I see in hardware : /dev/serial/by-id/usb-PARADOX_PARADOX_307USB_a41193d0-if00-port0
I had SSH Terminal in my HA to modify /etc/udev/rules.d/paradox-serial.rules or /etc/udev/rules.dev/paradox-serial.rules or /etc/udev/rules.d/paradox.rules and so on.
But nothing change, not working after reboot.

Any other suggestion to use this PAI ?

PS : My host machine (Win 10) is linked to EVO192 panel with Paradox USB307 serial cable
I did a try in python with pyserial (under my Win10) and I can see some of the event of the detector...
